conf: add docker-compose and pure-docker deploy types

ggilmore opened the pull request:
@dadlerj  I reverted https://github.com/sourcegraph/deploy-sourcegraph-docker/commit/f75a88c0a1708a3c5e26ea24b16ed191b2f1f424 because I didn't realize that new values of the `DEPLOY_TYPE` environment variable needed to be explicitly added to our internal configuration package. Before this revert, the frontend container would crash with the following error message:

```
2020/02/28 19:57:53 The 'DEPLOY_TYPE' environment variable is invalid. Expected one of: "cluster", "docker-container", "dev". Got: "docker-compose"
```
---

This PR adds support for `DEPLOY_TYPE=docker-compose` and `DEPLOY_TYPE=pure-docker` so that instances are properly classified for analytics and support purposes. 

A new release (3.13.2) will need to be cut for this.

uwedeportivo commented on internal/conf/computed.go:
```diff
@@ -136,9 +136,11 @@ func CanReadEmail() bool {
+	DeployCluster       = "cluster"
+	DeploySingleDocker  = "docker-container"
+	DeployDockerCompose = "docker-compose"
+	DeployPureDocker    = "pure-docker"
```
i didn't know about pure docker. we already have that ?

ggilmore replied:
Yes, that's what all the bash scripts in https://github.com/sourcegraph/deploy-sourcegraph-docker/ are for.

uwedeportivo replied:
ah, ok. thanks

slimsag commented:
Alternative to solving this:

Today this code handles _deployment type checking_, for when there need to be actual differences in our code with how a deployment type is handled. This is always one of three categories:

- Dev environment ("dev")
- Single-container deployment ("docker-container")
- Multi-container deployment ("Cluster")

The fact that we use this in tracking today is already bad because it conflates the above with what we're trying to track: the underlying deployment metadata. Instead I propose adding a new environment variable `DEPLOYMENT_META='type-k8s'` to the frontends only which we can set to an arbitrary tags and it will be passed through via the pings.

This way, we do not conflate deployment type checking and metadata collection about the delpoyment, and adding more metadata in the future about the deployment is easy if we need to do so.

ggilmore commented:
I agree that this solution is kludgey, but I actually _do_ think we need the ability to type switch on the specific platform in some places in our code base:

- https://github.com/sourcegraph/sourcegraph/tree/master/cmd/frontend/internal/app/pkg/updatecheck
- https://github.com/sourcegraph/sourcegraph/blob/master/web/src/repo/DirectImportRepoAlert.tsx

So, perhaps we should do the following:

- have a `DEPLOY_META` environment variable that contains arbitrary keys for pings
- have specific `conf.DeployTypes` (for when we need platform specific logic):
   - docker-compose
   - pure-docker
   - k8s
   - single docker image 
   - local development
- refactor most of the other call sites in https://sourcegraph.com/search?q=DeployType+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24&patternType=literal#13 to call methods that describes _traits_ of the deployment platform instead of hard-coding specific platforms when it's  it's not necessary. For example, https://sourcegraph.com/github.com/sourcegraph/sourcegraph@a0c222a/-/blob/web/src/search/results/SearchResultsList.tsx#L416 is really just checking to see if the current instance is running on something that doesn't have resource isolation. Perhaps we can have a `conf.HasResourceIsolation(deployType)` method that returns this information instead of checking for a specific string.

slimsag commented:
That approach SGTM

On Fri, Feb 28, 2020 at 3:03 PM Geoffrey Gilmore <notifications@github.com>
wrote:

> I agree that this solution is kludgey, but I actually *do* think we need
> the ability to type switch on the specific platform in some places in our
> code base:
>
>    -
>    https://github.com/sourcegraph/sourcegraph/tree/master/cmd/frontend/internal/app/pkg/updatecheck
>    -
>    https://github.com/sourcegraph/sourcegraph/blob/master/web/src/repo/DirectImportRepoAlert.tsx
>
> So, perhaps we should do the following:
>
>    - have a DEPLOY_META environment variable that contains arbitrary keys
>    for pings
>    - have specific conf.DeployTypes (for when we need platform specific
>    logic):
>       - docker-compose
>       - pure-docker
>       - k8s
>       - single docker image
>       - local development
>    - refactor most of the other call sites in
>    https://sourcegraph.com/search?q=DeployType+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24&patternType=literal#13
>    to call methods that describes *traits* of the deployment platform
>    instead of hard-coding specific platforms when it's it's not necessary. For
>    example,
>    https://sourcegraph.com/github.com/sourcegraph/sourcegraph@a0c222a/-/blob/web/src/search/results/SearchResultsList.tsx#L416
>    is really just checking to see if the current instance is running on
>    something that doesn't have resource isolation. Perhaps we can have a
>    conf.HasResourceIsolation(deployType) method that returns this
>    information instead of checking for a specific string.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/sourcegraph/sourcegraph/pull/8658?email_source=notifications&email_token=AAYGWOBFMXZ2DK5O4HQWIGTRFGC3XA5CNFSM4K54ARTKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENKKJYY#issuecomment-592749795>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAYGWOG3FKU3KI7WK43MAXTRFGC3XANCNFSM4K54ARTA>
> .
>
-- 
Follow me on twitter @slimsag <https://twitter.com/slimsag>.

keegancsmith commented:
This solution seems good to unblock tracking of deploy types in the mean time. The agreed upon solution above sounds good, but I wouldn't want to not track this information for another release if we don't find the time to work on it.

ggilmore commented:
@sourcegraph/distribution  @web PTAL, 

I have updated this PR with the following:

- https://github.com/sourcegraph/sourcegraph/blob/deploy-type-docker-compose/cmd/frontend/internal/app/pkg/updatecheck/handler.go , https://github.com/sourcegraph/sourcegraph/blob/deploy-type-docker-compose/cmd/frontend/internal/app/pkg/updatecheck/handler_test.go: Add a separate `latestReleaseDockerComposeOrPureDocker` version number for update checks. Existing Docker Compose / Pure Docker deployments will behave as they did before (looking at the latest Kubernetes version) since they currently identify as `cluster`. 

- https://github.com/sourcegraph/sourcegraph/blob/deploy-type-docker-compose/web/src/search/results/SearchResultsList.tsx and https://github.com/sourcegraph/sourcegraph/blob/deploy-type-docker-compose/web/src/repo/DirectImportRepoAlert.tsx: Added proper support for the new `docker-compose` and `pure-docker` deploy types by inverting the existing checks and explicitly stating the deploy types that the logic should apply to.

ggilmore commented:
I also don't understand why codecov is failing here?

unknwon commented:
> I also don't understand why codecov is failing here?

Trying my best, haha. I guess the diff in the PR used to have 40.93% coverage, now only has 39.13%.

felixfbecker commented:
This is weird, those commit statuses should be disabled: https://github.com/sourcegraph/sourcegraph/blob/bd8b3448dd5669bef7607a0436e1099ea00ec1b1/dev/codecov.yml#L5

slimsag commented on internal/conf/computed.go:
```diff
@@ -152,7 +154,7 @@ func DeployType() string {
 // IsDeployTypeCluster tells if the given deployment type is a cluster (and
-// non-dev, non-single Docker image).
+// non-dev, not docker-compose, not pure-docker, and non-single Docker image).
 func IsDeployTypeCluster(deployType string) bool {
```
"pure docker" and "docker compose" can be cluster deployments in the future, so I think calling this "is a cluster" is misleading if the intent is to signal Kubernetes. We should rename this to `IsDeployTypeKubernetes` if that is the intent here and you do not want multi-machine docker-compose deployments having this deploy type

This would also resolve confusion like this: https://sourcegraph.slack.com/archives/C07KZF47K/p1583516232089200?thread_ts=1583515677.088300&cid=C07KZF47K

slimsag commented on internal/conf/computed.go:
```diff
@@ -148,19 +148,19 @@ func DeployType() string {
+// IsDeployTypeKubernetes tells if the given deployment type is a Kubernetes
+// cluster (and non-dev, not docker-compose, not pure-docker, and non-single Docker image).
+func IsDeployTypeKubernetes(deployType string) bool {
+	if deployType == "cluster" {
```
```suggestion
	if deployType == "cluster" || deployType == "k8s" {
```

slimsag commented on internal/conf/computed.go:
```diff
@@ -136,7 +136,7 @@ func CanReadEmail() bool {
 // https://sourcegraph.com/search?q=r:github.com/sourcegraph/sourcegraph%24+%22type+DeployType%22
 const (
-	DeployCluster       = "cluster"
+	DeployKubernetes    = "cluster"
```
```suggestion
	DeployKubernetes    = "kubernetes"
```

I assume we'd want this otherwise these would show up in pings as "cluster" too?

slimsag commented on internal/conf/computed.go:
```diff
@@ -29,8 +29,8 @@ func defaultConfigForDeployment() conftypes.RawUnified {
-	case IsDeployTypeCluster(deployType), IsDeployTypeDockerCompose(deployType), IsDeployTypePureDocker(deployType):
-		return confdefaults.Cluster
+	case IsDeployTypeKubernetes(deployType), IsDeployTypeDockerCompose(deployType), IsDeployTypePureDocker(deployType):
+		return confdefaults.Kubernetes
```
```suggestion
		return confdefaults.KubernetesOrDockerComposeOrPureDocker
```

slimsag commented on internal/conf/confdefaults/confdefaults.go:
```diff
@@ -52,11 +52,11 @@ var DockerContainer = conftypes.RawUnified{
-var Cluster = conftypes.RawUnified{
+// Kubernetes is the default configuration applied to Kubernetes cluster
+// instances of Sourcegraph.
+var Kubernetes = conftypes.RawUnified{
```
```suggestion
var KubernetesOrDockerComposeOrPureDocker = conftypes.RawUnified{
```

slimsag commented on internal/conf/confdefaults/confdefaults.go:
```diff
@@ -52,11 +52,11 @@ var DockerContainer = conftypes.RawUnified{
-// Cluster is the default configuration applied to Cluster instances of
-// Sourcegraph.
-var Cluster = conftypes.RawUnified{
+// Kubernetes is the default configuration applied to Kubernetes cluster
```
```suggestion
// KubernetesOrDockerComposeOrPureDocker is the default configuration applied to Kubernetes,
// docker-compose, and pure-docker deployments of Sourcegraph.
```

slimsag commented on internal/conf/confdefaults/confdefaults.go:
```diff
@@ -52,11 +52,11 @@ var DockerContainer = conftypes.RawUnified{
-// Sourcegraph.
-var Cluster = conftypes.RawUnified{
+// Kubernetes is the default configuration applied to Kubernetes cluster
+// instances of Sourcegraph.
```
```suggestion
```

slimsag commented on internal/conf/validate.go:
```diff
@@ -273,7 +273,7 @@ func (f jsonLoaderFactory) New(source string) gojsonschema.JSONLoader {
 	mustValidate("DevAndTesting", confdefaults.DevAndTesting)
 	mustValidate("DockerContainer", confdefaults.DockerContainer)
-	mustValidate("Cluster", confdefaults.Cluster)
+	mustValidate("Kubernetes", confdefaults.Kubernetes)
```
```suggestion
	mustValidate("KubernetesOrDockerComposeOrPureDocker", confdefaults.KubernetesOrDockerComposeOrPureDocker)
```

slimsag commented on internal/conf/computed_test.go:
```diff
@@ -64,7 +64,7 @@ func TestSearchIndexEnabled(t *testing.T) {
 
 	defaults := map[string]conftypes.RawUnified{
-		"Cluster":         confdefaults.Cluster,
+		"Kubernetes":      confdefaults.Kubernetes,
```
```suggestion
		"KubernetesOrDockerComposeOrPureDocker":      confdefaults.KubernetesOrDockerComposeOrPureDocker,
```

slimsag commented on internal/conf/computed.go:
```diff
@@ -136,7 +136,7 @@ func CanReadEmail() bool {
 // Deploy type constants. Any changes here should be reflected in the DeployType type declared in web/src/globals.d.ts:
 // https://sourcegraph.com/search?q=r:github.com/sourcegraph/sourcegraph%24+%22type+DeployType%22
```
According to this comment you need to update globals.d.ts

ggilmore replied:
In what way? I already added the new docker-compose and pure-docker deploy types to that file.

ggilmore replied:
Oh, I see the rest of your changes now.

slimsag replied:
Ah weird, I searched but didn't find changes to that file in your PR -- I must've been in the review mode that selects a single commit.
